server/build_event_protocol/event_parser/event_parser.go

```diff
@@ -248,7 +249,13 @@ func FillInvocationFromEvents(buildEvents []*inpb.InvocationEvent, invocation *i
 		fillInvocationFromBuildMetadata(buildMetadatum, invocation)
 	}
 
-	buildDuration := time.Duration((endTimeMillis - startTimeMillis) * int64(time.Millisecond))
+	buildDuration := time.Duration(int64(0))
+	if startTimeMillis != undefinedTimestamp {
+		if endTimeMillis == undefinedTimestamp {
+			endTimeMillis = time.Now().UnixNano() / time.Millisecond.Nanoseconds()
+		}
+		buildDuration = time.Duration((endTimeMillis - startTimeMillis) * int64(time.Millisecond))
+	}
```

Comment:
is startTimeMillis ever undefined? If possible it'd be great to simplify to just:

if endTimeMillis == undefinedTimestamp {
  duration = time.Since(startTime)
} else {
  duration = endTime - startTime
}

Reply:
Could initialize the start time to Now() instead of -1 to make sure it’s never not set

Reply:
I like the idea of initializing both `startTime` and `endTime` to `Now()`; makes the implementation very clean. The only "weird" case that might happen is that `startTimeMillis` might be undefined but `endTimeMillis` is defined, then we get a negative duration. But it doesn't seem like it's worth worrying about that case since the `buildEvent.started` (which has the start time) is "normally" the first event (according to the comments in the proto)

Reply:
Ah yeah, I guess you could still get negative durations. I also just realized this is in FillInvocationFromEvents so it will likely be the time that LookupInvocation is called, not the time FinalizeInvocation is called (when calling GetInvocation at least).

I'd test this out a bit before checking something in here. User clocks being different than the server clock (not timezone, just set to a totally different time) could also lead to negative durations here.